readme.md
<p align="center">
  <img width="128" height="128" src="https://user-images.githubusercontent.com/2746374/179428323-6d6b5e7f-6459-47b3-9e03-fea7553d85af.svg" />
</p>

<h1 align="center">Nightfox</h1>

<p align="center">
  <a href="https://github.com/edeneast/nightfox.nvim/stargazers"><img src="https://img.shields.io/github/stars/edeneast/nightfox.nvim?colorA=192330&colorB=719cd6&style=for-the-badge"></a>
  <a href="https://github.com/edeneast/nightfox.nvim/issues"><img src="https://img.shields.io/github/issues/edeneast/nightfox.nvim?colorA=192330&colorB=dbc074&style=for-the-badge"></a>
  <a href="https://github.com/edeneast/nightfox.nvim/contributors"><img src="https://img.shields.io/github/contributors/edeneast/nightfox.nvim?colorA=192330&colorB=81b29a&style=for-the-badge"></a>
  <a href="https://github.com/EdenEast/nightfox.nvim/network/members"><img src="https://img.shields.io/github/forks/edeneast/nightfox.nvim?colorA=192330&colorB=9d79d6&style=for-the-badge"></a>
</p>

<p align="center">
  A highly customizable theme for vim and neovim with support for lsp, treesitter and a variety of plugins.
</p>

<div align="center">
  <h3>Nightfox</h3><img src="https://user-images.githubusercontent.com/2746374/158456286-9e3ee657-60e6-49d8-b85e-dcab285b31c3.png" alt="nightfox" style="border-radius:1%" />
  <h3>Dayfox</h3><img src="https://user-images.githubusercontent.com/2746374/210672782-6b8690d0-3ef5-4f32-bdea-4f0a97b9d9d5.png" alt="dayfox" style="border-radius:1%" />
  <h3>Dawnfox</h3><img src="https://user-images.githubusercontent.com/2746374/158456278-c5d656de-c445-44b8-9813-9fc91ffbce4c.png" alt="dawnfox" style="border-radius:1%" />
  <h3>Duskfox</h3><img src="https://user-images.githubusercontent.com/2746374/158456284-b6ff583f-c0bb-417c-b2d5-60b8364049a7.png" alt="duskfox" style="border-radius:1%" />
  <h3>Nordfox</h3><img src="https://user-images.githubusercontent.com/2746374/158456289-eee096df-7ea7-4161-9199-808fdba58e06.png" alt="nordfox" style="border-radius:1%" />
  <h3>Terafox</h3><img src="https://user-images.githubusercontent.com/2746374/160268126-cd417690-3660-4a04-9040-183c35f127f7.png" alt="terafox" style="border-radius:1%" />
  <h3>Carbonfox</h3><img src="https://user-images.githubusercontent.com/2746374/183569670-91260cab-83a9-4247-8e09-6abb46c52152.png" alt="carbonfox" style="border-radius:1%" />
</div>

> Tabline is [tabby.nvim] and statusline is [feline.nvim]. You can find my config for [tabby] and [feline] here as well
> as the auto generated [highlight][conf-colors] groups here. If you would like a single consumable file that contains
> that can be added to your own config check out [misc/feline.lua][misc-feline] and [misc/tabby.lua][misc-tabby].

[tabby.nvim]: https://github.com/nanozuki/tabby.nvim
[feline.nvim]: https://github.com/freddiehaddad/feline.nvim
[tabby]: https://github.com/EdenEast/nyx/blob/573ff3f22da95c00ec71a094b3e1ec166c0afcdd/config/.config/nvim/lua/eden/modules/ui/tabby.lua
[feline]: https://github.com/EdenEast/nyx/blob/573ff3f22da95c00ec71a094b3e1ec166c0afcdd/config/.config/nvim/lua/eden/modules/ui/feline/init.lua
[conf-colors]: https://github.com/EdenEast/nyx/blob/573ff3f22da95c00ec71a094b3e1ec166c0afcdd/config/.config/nvim/lua/eden/modules/ui/colors.lua
[misc-feline]: ./misc/feline.lua
[misc-tabby]: ./misc/tabby.lua

## Features

- Supports both vim and neovim
- Highly configurable with template overriding
- [Colorblind](#colorblind) mode (daltonization, and simulation)
- Support for multiple [plugins](#supported-plugins) and [status lines](#status-lines)
  - And many others should "just work"!
- [Compile](#compile) user's configuration for fast startup times
- Export [Color](#color-lib) library utility
- [Interactive](#interactive) live config re-loading

## Requirements

- Neovim >= 0.8 **or** Vim 9 with lua = **5.1+**
- True color support
- Undercurl terminal support (optional)

> Note for MacOS: The default `terminal` does not support `true color`. Use either `Iterm2` or [another] terminal that
> supports true color.

[another]: https://gist.github.com/XVilka/8346728#terminal-emulators

## Installation

Download with your favorite package manager.

```lua
{ "EdenEast/nightfox.nvim" } -- lazy
```

```lua
use "EdenEast/nightfox.nvim" -- Packer
```

```vim
Plug 'EdenEast/nightfox.nvim' " Vim-Plug
```

## Usage

Simply set the colorscheme with the builtin command `:colorscheme`

```vim
colorscheme nightfox
```

```lua
vim.cmd("colorscheme nightfox")
```

## Configuration

There is no need to call `setup` if you don't want to change the default options and settings.

```lua
-- Default options
require('nightfox').setup({
  options = {
    -- Compiled file's destination location
    compile_path = vim.fn.stdpath("cache") .. "/nightfox",
    compile_file_suffix = "_compiled", -- Compiled file suffix
    transparent = false,     -- Disable setting background
    terminal_colors = true,  -- Set terminal colors (vim.g.terminal_color_*) used in `:terminal`
    dim_inactive = false,    -- Non focused panes set to alternative background
    module_default = true,   -- Default enable value for modules
    colorblind = {
      enable = false,        -- Enable colorblind support
      simulate_only = false, -- Only show simulated colorblind colors and not diff shifted
      severity = {
        protan = 0,          -- Severity [0,1] for protan (red)
        deutan = 0,          -- Severity [0,1] for deutan (green)
        tritan = 0,          -- Severity [0,1] for tritan (blue)
      },
    },
    styles = {               -- Style to be applied to different syntax groups
      comments = "NONE",     -- Value is any valid attr-list value `:help attr-list`
      conditionals = "NONE",
      constants = "NONE",
      functions = "NONE",
      keywords = "NONE",
      numbers = "NONE",
      operators = "NONE",
      strings = "NONE",
      types = "NONE",
      variables = "NONE",
    },
    inverse = {             -- Inverse highlight for different types
      match_paren = false,
      visual = false,
      search = false,
    },
    modules = {             -- List of various plugins and additional options
      -- ...
    },
  },
  palettes = {},
  specs = {},
  groups = {},
})

-- setup must be called before loading
vim.cmd("colorscheme nightfox")
```

If you would like to change any of the default options above you only have to define the options that change. If an
option is not present in your options table the default option will be used. For example if changing the styles of
certain syntax is the only desired change then your options table would look like:

```lua
require('nightfox').setup({
  options = {
    styles = {
      comments = "italic",
      keywords = "bold",
      types = "italic,bold",
    }
  }
})
```

To see a detailed explanation of nightfox's config settings consult either the builtin help `:help nightfox` or
[Usage](./usage.md).

### Modules

Nightfox's modules store configuration information for various plugins and other neovim modules. To enable a module
either set the module to `true` or if the module has additional configuration information set `enable` to `true`.

By default modules will be enabled. To change this behaviour change `options.module_default` to `false`.

To see a list of modules check, see [Usage](./usage.md#module) or `:help` for more information.

### Customize palettes and groups

You can change the color `palette` and the highlight `group` of nightfox. Here is a brief example:

```lua
-- Palettes are the base color defines of a colorscheme.
-- You can override these palettes for each colorscheme defined by nightfox.
local palettes = {
  -- Everything defined under `all` will be applied to each style.
  all = {
    -- Each palette defines these colors:
    --   black, red, green, yellow, blue, magenta, cyan, white, orange, pink
    --
    -- These colors have 3 shades: base, bright, and dim
    --
    -- Defining just a color defines it's base color
    red = "#ff0000",
  },
  nightfox = {
    -- A specific style's value will be used over the `all`'s value
    red = "#c94f6d",
  },
  dayfox = {
    -- Defining multiple shades is done by passing a table
    blue = { base = "#4d688e", bright = "#4e75aa", dim = "#485e7d" },
  },
  nordfox = {
    -- A palette also defines the following:
    --   bg0, bg1, bg2, bg3, bg4, fg0, fg1, fg2, fg3, sel0, sel1, comment
    --
    -- These are the different foreground and background shades used by the theme.
    -- The base bg and fg is 1, 0 is normally the dark alternative. The others are
    -- incrementally lighter versions.
    bg1 = "#2e3440",

    -- sel is different types of selection colors.
    sel0 = "#3e4a5b", -- Popup bg, visual selection bg
    sel1 = "#4f6074", -- Popup sel bg, search bg

    -- comment is the definition of the comment color.
    comment = "#60728a",
  },
}

-- Spec's (specifications) are a mapping of palettes to logical groups that will be
-- used by the groups. Some examples of the groups that specs map would be:
--   - syntax groups (functions, types, keywords, ...)
--   - diagnostic groups (error, warning, info, hints)
--   - git groups (add, removed, changed)
--
-- You can override these just like palettes
local specs = {
  -- As with palettes, the values defined under `all` will be applied to every style.
  all = {
    syntax = {
      -- Specs allow you to define a value using either a color or template. If the string does
      -- start with `#` the string will be used as the path of the palette table. Defining just
      -- a color uses the base version of that color.
      keyword = "magenta",

      -- Adding either `.bright` or `.dim` will change the value
      conditional = "magenta.bright",
      number = "orange.dim",
    },
    git = {
      -- A color define can also be used
      changed = "#f4a261",
    },
  },
  nightfox = {
    syntax = {
      -- As with palettes, a specific style's value will be used over the `all`'s value.
      operator = "orange",
    },
  },
}

-- Groups are the highlight group definitions. The keys of this table are the name of the highlight
-- groups that will be overridden. The value is a table with the following values:
--   - fg, bg, style, sp, link,
--
-- Just like `spec` groups support templates. This time the template is based on a spec object.
local groups = {
  -- As with specs and palettes, the values defined under `all` will be applied to every style.
  all = {
    -- If `link` is defined it will be applied over any other values defined
    Whitespace = { link = "Comment" },

    -- Specs are used for the template. Specs have their palette's as a field that can be accessed
    IncSearch = { bg = "palette.cyan" },
  },
  nightfox = {
    -- As with specs and palettes, a specific style's value will be used over the `all`'s value.
    PmenuSel = { bg = "#73daca", fg = "bg0" },
  },
}

require("nightfox").setup({ palettes = palettes, specs = specs, groups = groups })

-- setup must be called before loading
vim.cmd("colorscheme nightfox")
```

To find the list of syntax highlight groups defined for vim use the help `:help group-name` and
`:help nvim-treesitter-highlights` for treesitter. If you would also like to see how nightfox defines these highlight groups
you can see [syntax.lua] for vim's syntax and [treesitter.lua] for treesitter. These files list out all all highlight
groups and have a comment describing them. Another file to note is [editor.lua] which is the highlight groups responsible
for how vim looks (background, cursorline, tabline, etc...). To get the highlight group under your cursor see
[here](#syntax-highlight-groups) for more information

To get more information check out [Usage](./usage.md#configuration) or the help file `:help nightfox` for more detailed information.

[editor.lua]: https://github.com/EdenEast/nightfox.nvim/blob/main/lua/nightfox/group/editor.lua
[syntax.lua]: https://github.com/EdenEast/nightfox.nvim/blob/main/lua/nightfox/group/syntax.lua
[treesitter.lua]: https://github.com/EdenEast/nightfox.nvim/blob/main/lua/nightfox/group/modules/treesitter.lua

### Custom template values

Nightfox's `palettes` and `specs` can be extended with your own values. This is useful for users that want to
distinguish a spec value being used for multiple group definitions. This is best understood with an example:

`bg0` is used as the dark alternative background color. This is used in multiple areas (non-current file, status line,
normal float, etc...). If you would like to have the inactive color diffferent then the float / statusline color you can
define your own value in the `specs` table.

```lua
require("nightfox").setup({
  palettes = {
    -- Custom duskfox with black background
    duskfox = {
      bg1 = "#000000", -- Black background
      bg0 = "#1d1d2b", -- Alt backgrounds (floats, statusline, ...)
      bg3 = "#121820", -- 55% darkened from stock
      sel0 = "#131b24", -- 55% darkened from stock
    },
  },
  specs = {
    all = {
      inactive = "bg0", -- Default value for other styles
    },
    duskfox = {
      inactive = "#090909", -- Slightly lighter then black background
    },
  },
  groups = {
    all = {
      NormalNC = { fg = "fg1", bg = "inactive" }, -- Non-current windows
    },
  },
})
```

## Api

Nightfox exposes some Api's and utility classes that let you fetch data from nightfox.

### Palettes

You can get the palettes used by each colorscheme:

```lua
-- Returns a table with each colorscheme and the palette associated with it
local palettes = require('nightfox.palette').load()

-- Returns the palette of the specified colorscheme
local palette = require('nightfox.palette').load("nightfox")

print(vim.inspect(palette.red))
-- {
--   base = "#c94f6d",
--   bright = "#d16983",
--   dim = "#ab435d",
-- }
```

See [Usage](./usage.md#palette) for more information on palettes.

### Specs

You can get the spec used by each colorscheme:

```lua
-- Returns a table with each colorscheme and the spec associated with it
local specs require('nightfox.spec').load()

-- Returns the spec of the specified colorscheme
local spec = require('nightfox.spec').load("nightfox")

print(vim.inspect(spec.git))
-- {
--   add = "#81b29a",
--   changed = "#dbc074",
--   removed = "#c94f6d"
-- }
```

See [Usage](./usage.md#spec) for more information on specs.

### Color lib

Nightfox uses a color library internally to manipulate colors. You can use this library as well.

```lua
local palette = require('nightfox.palette').load('nightfox')
local Color = require("nightfox.lib.color")

local bg = Color.from_hex(palette.bg1)
local red = Color.from_hex("#ff0000")

-- Blend the bg with red. The blend factor is from 0 to 1
-- with 0 being full bg and 1 being full red
local red_bg = bg:blend(red, 0.2)

print(red_bg:to_css())
-- "#471c26"

-- Brighten bg by adding 10 to the value of the color as a hsv
local alt_bg = bg:brighten(10)
print(vim.inspect(alt_bg:to_hsv()))
-- {
--   hue = 213.91304347826,
--   saturation = 47.916666666667,
--   value = 28.823529411765
-- }
```

There are a lot of useful functions to manipulate and work with colors in different color spaces.
See [Usage](./usage.md#color) for more information on `Color`.

## Colorblind

For individuals with `color vision deficiency` (cvd), nightfox has implemented a `colorblind` mode to help enhance color
contrast. This can be enabled with this option `colorblind.enable`.

<details>
<summary>Understanding cvd</summary>

There are three types of cvd:

- Protan (Red / L cones)
- Deutan (Green / M cones)
- Tritan (Blue / S cones)

These are referred to as `protanomaly`, `deuteranomaly`, and `tritanomaly` for individuals that have all three cones
(trichromats) but one is week (anomalous trichromacy).

These can also be referred to as `protanopia`, `deuteranopia`, and `tritanopia`. This is for individuals that only have
two cones (dichromats or dichromacy).

| Cone      | Type   | Week (trichromacy) | Missing (Dichromacy) |
| --------- | ------ | ------------------ | -------------------- |
| L / Red   | Protan | Protanomaly        | Protanopia           |
| M / Green | Deutan | Deuteranomaly      | Deuteranopia         |
| S / Blue  | Tritan | Tritanomaly        | Tritanopia           |

</details>

### Configuring cvd

Nightfox needs to simulate your cvd in order to shift colors correctly. This is done by setting your cvd type's severity
level. Severity is a value between `0` and `1` where `1` is full dichromacy. You can also have multiple kinds of cvd
configured at a time. Here is a full example:

```lua
require("nightfox").setup({
  options = {
    colorblind = {
      enable = true,
      severity = {
        protan = 0.3,
        deutan = 0.6,
      },
    },
  },
})
```

If you are looking for a way to self evaluate what severity factor to use, check out [daltonlens's][cb-self-eval] self
evaluation article with interactive self evaluation Ishihasa plates.

Another method would be to use the option `colorblind.simulate_only` option along with nightfox's
[interactive](#interactive) mode. While nightfox is simulating cvd set a severity to 1. Now decrease the severity
incrementally until you cannot perceive a difference in the change of colors.

[cb-self-eval]: https://daltonlens.org/evaluating-cvd-simulation/#Generating-Ishihara-like-plates-for-self-evaluation

### How does this work?

This is accomplished by applying an algorithm called `Daltonization`. The process follows these steps:

1. Simulate what a person with cvd would see
1. Calculate the difference between original vs. simulated
1. Shift the difference towards the visible spectrum of the cvd individual
1. Correct original color by adding it to the corrected difference

You can see the simulated colors instead of the corrected colors by setting the option `colorblind.simulate_only`.

![cvd-example](https://user-images.githubusercontent.com/2746374/210025850-9a84b142-e989-4efa-9b55-5f7312013da3.gif)

## Compile

Nightfox is a highly customizable and configurable colorscheme. This does however come at the cost of complexity and
execution time.

Nightfox pre-computes the result of your configuration and saves the lua bytecode in a cache to be used on next load.
This significantly speeds up nightfox's execution time. Changes to your configuration will be re-computed and cached
automatically.

By default nightfox writes the compiled results into the system's `cache` directory. On unix this is
`$XDG_CACHE_HOME/nvim/nightfox` and on windows this is `%localappdata%\\Temp\\nvim\\nightfox`.

Nightfox provides functions to work with the nightfox compiler.

```vim
:NightfoxCompile " Manually call nightfox compiler to create/update compiled files
```

```lua
require('nightfox').compile() -- lua api version
```

## Interactive

Nightfox makes it easy to make changes to its config and see the results. For this nightfox exposes the command:

```
NightfoxInteractive
```

This command will attach an autocmd to the current buffer that executes on `BufferWritePost`. The
autocmd will clear nightfox's internal state and re-source it's config from the newly saved file. It will then reset the
colorscheme.

![demo](https://user-images.githubusercontent.com/2746374/211108157-b5221ea4-9369-453d-9c27-4a71ac161a55.gif)

There are a few things to note:

- This requires executing `luafile` on the current file. Any syntax errors will throw errors.
- If you are using packer and have nightfox's config in a `config = function() end` block, this will not work as packer
  would require to be re-compiled and the compiled file sourced.

## Supported Plugins

- [alpha-nvim](https://github.com/goolord/alpha-nvim)
- [aerial.nvim](https://github.com/stevearc/aerial.nvim)
- [barbar.nvim](https://github.com/romgrk/barbar.nvim)
- [nvim-cmp](https://github.com/hrsh7th/nvim-cmp)
- [nvim-dap-ui](https://github.com/rcarriga/nvim-dap-ui)
- [dashboard-nvim](https://github.com/glepnir/dashboard-nvim)
- [fern.vim](https://github.com/lambdalisue/fern.vim)
- [fidget.nvim](https://github.com/j-hui/fidget.nvim)
- [vim-gitgutter](https://github.com/airblade/vim-gitgutter)
- [gitsigns.nvim](https://github.com/lewis6991/gitsigns.nvim)
- [glyph-palette.vim](https://github.com/lambdalisue/glyph-palette.vim)
- [hop.nvim](https://github.com/phaazon/hop.nvim)
- [vim-illuminate](https://github.com/RRethy/vim-illuminate)
- [lightspeed.nvim](https://github.com/ggandor/lightspeed.nvim)
- [lspsaga.nvim](https://github.com/glepnir/lspsaga.nvim)
- [lsp-trouble.nvim](https://github.com/simrat39/lsp-trouble.nvim)
- [indent-blankline.nvim](https://github.com/lukas-reineke/indent-blankline.nvim)
- [mini.nvim](https://github.com/echasnovski/mini.nvim)
- [modes.nvim](https://github.com/mvllow/modes.nvim)
- [nvim-navic](https://github.com/SmiteshP/nvim-navic)
- [neogit](https://github.com/TimUntersberger/neogit)
- [neotest](https://github.com/nvim-neotest/neotest)
- [neo-tree](https://github.com/nvim-neo-tree/neo-tree.nvim)
- [nvim-notify](https://github.com/rcarriga/nvim-notify)
- [nvim-tree.lua](https://github.com/kyazdani42/nvim-tree.lua)
- [pounce.nvim](https://github.com/rlane/pounce.nvim)
- [vim-signify](https://github.com/mhinz/vim-signify)
- [vim-sneak](https://github.com/justinmk/vim-sneak)
- [symbols-outline.nvim](https://github.com/simrat39/symbols-outline.nvim)
- [telescope.nvim](https://github.com/nvim-telescope/telescope.nvim)
- [nvim-ts-rainbow](https://github.com/p00f/nvim-ts-rainbow)
- [nvim-ts-rainbow2](https://github.com/HiPhish/nvim-ts-rainbow2)
- [which-key.nvim](https://github.com/folke/which-key.nvim)

## Status lines

### [Lualine]

Lualine checks the value of `vim.g.colors_name` (set when using `:colorscheme` command) to determine the theme to load.
Set your colorscheme before calling setup.

```lua
vim.cmd("colorscheme nightfox")
require('lualine').setup({ ... })
```

### [Lightline]

Lightline checks the value of `vim.g.lightline.colorscheme`. It does not depend on
the value of the `:colorscheme` command, Please set it.

```lua
vim.cmd("colorscheme nightfox")
vim.g.lightline = { colorscheme = "nightfox" } -- Or the name of colorscheme you use
```

## Syntax highlight groups

This section will help you determine what highlight group is being applied to a piece of syntax. These sections will
output the highlight group for the value under the cursor.

#### Treesitter highlighting


CHANGELOG.md
# Changelog

## [4.0.0](https://github.com/Swoogan/nightfox.nvim/compare/v3.6.1...v4.0.0) (2023-10-02)


### ⚠ BREAKING CHANGES

* **palette:** rework dayfox ([#285](https://github.com/Swoogan/nightfox.nvim/issues/285))
* **treesitter:** `TS*` highlight groups are removed.
* **group:** `groups` options now takes options per style + `all`

### Features

* Add `NightfoxLoad` command with completion ([c1ba098](https://github.com/Swoogan/nightfox.nvim/commit/c1ba0987f832fdc86035a74056df36ee93a9db73))
* Add `reset()` cmd to clear overrides and config ([5d2581a](https://github.com/Swoogan/nightfox.nvim/commit/5d2581a71510c319d128a8b02a21181abc611202))
* Add base dayfox theme ([62b2500](https://github.com/Swoogan/nightfox.nvim/commit/62b2500a187cfb2518ecd63eea04a343c16c7d3c))
* Add color override option to config ([1a61db5](https://github.com/Swoogan/nightfox.nvim/commit/1a61db502fccdc72a03f41b604cee59d5b92b90b))
* Add colorscheme for lualine and lightline ([3909e17](https://github.com/Swoogan/nightfox.nvim/commit/3909e177474459f8223ef6069e1e0b720011d5aa))
* add CurSearch highlight group ([#143](https://github.com/Swoogan/nightfox.nvim/issues/143)) ([de12896](https://github.com/Swoogan/nightfox.nvim/commit/de12896bc40461377adf9dfbabab89aad991f399))
* Add dayfox to extras ([069d15c](https://github.com/Swoogan/nightfox.nvim/commit/069d15c0bebc4a4d81619dc7a374a0c117878c72))
* Add hardfox colorscheme ([1bb7947](https://github.com/Swoogan/nightfox.nvim/commit/1bb79479ad52a263be11d6a59a2565ee55031bad))
* Add internal _colorscheme_load func ([d5a15a4](https://github.com/Swoogan/nightfox.nvim/commit/d5a15a4eb28bdbcae54a54b73197a1ce8ec2fa38))
* Add inverse selection options to config ([2b3ee99](https://github.com/Swoogan/nightfox.nvim/commit/2b3ee999e293933646f80855eb4f1b51fd496325)), closes [#31](https://github.com/Swoogan/nightfox.nvim/issues/31)
* Add light boolean value to colors object ([bd3b11f](https://github.com/Swoogan/nightfox.nvim/commit/bd3b11fd58a96b77551ec7a767f5827c7c2da435))
* Add name to color style ([c2a00bf](https://github.com/Swoogan/nightfox.nvim/commit/c2a00bff72216c88b0ee59fc82ca424a6d9a0ba3))
* add new DiagnosticOk highlight ([#378](https://github.com/Swoogan/nightfox.nvim/issues/378)) ([fe2fc7b](https://github.com/Swoogan/nightfox.nvim/commit/fe2fc7b93d66349eff2c5baa6cec922ee3958f56))
* add new lsp semantic token highlight groups ([#318](https://github.com/Swoogan/nightfox.nvim/issues/318)) ([a8044b0](https://github.com/Swoogan/nightfox.nvim/commit/a8044b084e0114609ec2c59cc4fa94c709a457d4))
* Add nordfox colorscheme style ([175d7df](https://github.com/Swoogan/nightfox.nvim/commit/175d7df35d4fcb11af26c5c45b857ec6e9dd0962))
* Add palefox ([ea4815b](https://github.com/Swoogan/nightfox.nvim/commit/ea4815b7bc9f1099362cc98a42cecfa2b0d7f7c8))
* Add plugin support ([25651a7](https://github.com/Swoogan/nightfox.nvim/commit/25651a7f10c3871d0bf330beed2064f4d9b4d778))
* Add randfox as a style option ([e2a53a7](https://github.com/Swoogan/nightfox.nvim/commit/e2a53a7e0f2e893f81272efc96e9157ab6a19f2e))
* Add string style config option ([1462a29](https://github.com/Swoogan/nightfox.nvim/commit/1462a298ae06f027669196c6055b1c31f88d4b9e))
* add support for ts-rainbow2 ([#301](https://github.com/Swoogan/nightfox.nvim/issues/301)) ([97e66df](https://github.com/Swoogan/nightfox.nvim/commit/97e66dfaaecdc81f1dd7da00f8d4a6033f3bf6b2))
* Add Telescope selection colors ([2466368](https://github.com/Swoogan/nightfox.nvim/commit/2466368383f65985c5135d9022329129080f9314))
* Add terminal color options ([2e177d2](https://github.com/Swoogan/nightfox.nvim/commit/2e177d21e0ca37affb94d04e29f96aad8581c58d))
* add vim/neovim test on ubuntu/macos/windows ([#320](https://github.com/Swoogan/nightfox.nvim/issues/320)) ([8bb6713](https://github.com/Swoogan/nightfox.nvim/commit/8bb6713c56458aae339575b205234d820ec2046a))
* Change TSField to blue ([13928d9](https://github.com/Swoogan/nightfox.nvim/commit/13928d9db2cfea7f77cb13e654a8e3ef526b779c))
* **ci:** add github action to autogenerate extra files ([#327](https://github.com/Swoogan/nightfox.nvim/issues/327)) ([727a0ea](https://github.com/Swoogan/nightfox.nvim/commit/727a0ea9d2dde7f8cfd4ae0f2e5570c365638227))
* **color:** add WCAG AA contrast validation ([12e0ca7](https://github.com/Swoogan/nightfox.nvim/commit/12e0ca70e978f58318e7f0279bb7b243ababbd49))
* **colors:** Add dawn and dusk colorschemes ([9762d29](https://github.com/Swoogan/nightfox.nvim/commit/9762d2939e0fd35b98f9115eb61e206f167cb2c4))
* **colors:** Warn when style not found ([1d423fa](https://github.com/Swoogan/nightfox.nvim/commit/1d423faa652907b3d9211ac1e44aee8adfbe210d))
* **compiler:** handle errors in compiled results ([#263](https://github.com/Swoogan/nightfox.nvim/issues/263)) ([e6a0d98](https://github.com/Swoogan/nightfox.nvim/commit/e6a0d98dc1b4751b4d999d3dffd87b1af4f4873e))
* **compiler:** write compile lua file if debug set ([f32eef4](https://github.com/Swoogan/nightfox.nvim/commit/f32eef468a98e73d5a87daa8db873dd1a94e021e))
* **compile:** simplify cache check ([#268](https://github.com/Swoogan/nightfox.nvim/issues/268)) ([e29a913](https://github.com/Swoogan/nightfox.nvim/commit/e29a91388d5a90109bf7c268782cd2301b7f3ed1))
* **config:** Add missing syntax styles ([#129](https://github.com/Swoogan/nightfox.nvim/issues/129)) ([a85405b](https://github.com/Swoogan/nightfox.nvim/commit/a85405b775a0a6a197556545359af0e9bcb13691))
* **config:** Default module config option ([#239](https://github.com/Swoogan/nightfox.nvim/issues/239)) ([0731d09](https://github.com/Swoogan/nightfox.nvim/commit/0731d0980afead7a6fffd55334b6bbe396801ab1))
* **cvd:** support colorblind mode with daltonization ([#281](https://github.com/Swoogan/nightfox.nvim/issues/281)) ([f093297](https://github.com/Swoogan/nightfox.nvim/commit/f093297145e917f7ae4d0e09fc9c07ac40620361))

lua/nightfox/util.lua
local hsluv = require("nightfox.hsluv")

local util = {}

util.bg = "#000000"
util.fg = "#ffffff"
util.day_brightness = 0.3

function util.hex_to_rgb(hex_str)
  local hex = "[abcdef0-9][abcdef0-9]"
  local pat = "^#(" .. hex .. ")(" .. hex .. ")(" .. hex .. ")$"
  hex_str = string.lower(hex_str)

  assert(string.find(hex_str, pat) ~= nil, "hex_to_rgb: invalid hex_str: " .. tostring(hex_str))

  local r, g, b = string.match(hex_str, pat)
  return { tonumber(r, 16), tonumber(g, 16), tonumber(b, 16) }
end

function util.hex_to_norm_rgb(hex_str)
  local c = util.hex_to_rgb(hex_str)
  return { c[1] / 255, c[2] / 255, c[3] / 255 }
end

---@param fg string foreground color
---@param bg string background color
---@param alpha number number between 0 and 1. 0 results in bg, 1 results in fg
function util.blend(fg, bg, alpha)
  bg = util.hex_to_rgb(bg)
  fg = util.hex_to_rgb(fg)

  local blendChannel = function(i)
    local ret = (alpha * fg[i] + ((1 - alpha) * bg[i]))
    return math.floor(math.min(math.max(0, ret), 255) + 0.5)
  end

  return string.format("#%02X%02X%02X", blendChannel(1), blendChannel(2), blendChannel(3))
end

function util.darken(hex, amount, bg)
  return util.blend(hex, bg or util.bg, math.abs(amount))
end

function util.lighten(hex, amount, fg)
  return util.blend(hex, fg or util.fg, math.abs(amount))
end

function util.brighten(color, percentage)
  local hsl = hsluv.hex_to_hsluv(color)
  local larpSpace = 100 - hsl[3]
  if percentage < 0 then
    larpSpace = hsl[3]
  end
  hsl[3] = hsl[3] + larpSpace * percentage
  return hsluv.hsluv_to_hex(hsl)
end

function util.invertColor(color)
  if color ~= "NONE" then
    local hsl = hsluv.hex_to_hsluv(color)
    hsl[3] = 100 - hsl[3]
    if hsl[3] < 40 then
      hsl[3] = hsl[3] + (100 - hsl[3]) * util.day_brightness
    end
    return hsluv.hsluv_to_hex(hsl)
  end
  return color
end

function util.color_overrides(colors, config)
  if type(config.colors) == "table" then
    for key, value in pairs(config.colors) do
      if not colors[key] then
        error("Color " .. key .. " does not exist")
      end

      -- Patch: https://github.com/ful1e5/onedark.nvim/issues/6
      if type(colors[key]) == "table" then
        util.color_overrides(colors[key], { colors = value })
      else
        if value:lower() == "none" then
          -- set to none
          colors[key] = "NONE"
        elseif string.sub(value, 1, 1) == "#" then
          -- hex override
          colors[key] = value
        else
          -- another group
          if not colors[value] then
            error("Color " .. value .. " does not exist")
          end
          colors[key] = colors[value]
        end
      end
    end
  end
end

function util.string_to_color(colors, value, default)
  if not value or value == "" then
    return default
  end

  -- If the value is a hex color code then return it
  local hex = "[abcdef0-9][abcdef0-9]"
  local pat = "^#" .. hex .. hex .. hex .. "$"
  if string.match(value, pat) then
    return value
  end

  local acceptable_colors = { "black", "red", "green", "blue", "magenta", "cyan", "white", "orange", "pink" }
  for _, ac in ipairs(acceptable_colors) do
    if string.match(value, ac) then
      return colors[value]
    end
  end

  -- Did not match anything to return default
  return default
end

function util.highlight(group, color)
  local style = color.style and "gui=" .. color.style or "gui=NONE"
  local fg = color.fg and "guifg=" .. color.fg or "guifg=NONE"
  local bg = color.bg and "guibg=" .. color.bg or "guibg=NONE"
  local sp = color.sp and "guisp=" .. color.sp or ""
  local hl = "highlight " .. group .. " " .. style .. " " .. fg .. " " .. bg .. " " .. sp

  vim.cmd(hl)
  if color.link then
    vim.cmd("highlight! link " .. group .. " " .. color.link)
  end
end

-- Simple string interpolation.
--
-- Example template: "${name} is ${value}"
--
---@param str string template string
---@param table table key value pairs to replace in the string
function util.template(str, table)
  return (str:gsub("($%b{})", function(w)
    return table[w:sub(3, -2)] or w
  end))
end

function util.syntax(tbl)
  for group, colors in pairs(tbl) do
    util.highlight(group, colors)
  end
end

function util.terminal(theme)
  vim.g.terminal_color_0 = theme.colors.black
  vim.g.terminal_color_1 = theme.colors.red
  vim.g.terminal_color_2 = theme.colors.green
  vim.g.terminal_color_3 = theme.colors.yellow
  vim.g.terminal_color_4 = theme.colors.blue
  vim.g.terminal_color_5 = theme.colors.magenta
  vim.g.terminal_color_6 = theme.colors.cyan
  vim.g.terminal_color_7 = theme.colors.white

  vim.g.terminal_color_8 = theme.colors.black_br
  vim.g.terminal_color_9 = theme.colors.red_br
  vim.g.terminal_color_10 = theme.colors.green_br
  vim.g.terminal_color_11 = theme.colors.yellow_br
  vim.g.terminal_color_12 = theme.colors.blue_br
  vim.g.terminal_color_13 = theme.colors.magenta_br
  vim.g.terminal_color_14 = theme.colors.cyan_br
  vim.g.terminal_color_15 = theme.colors.white_br
end

function util.load(theme)
  -- only needed to clear when not the default colorscheme
  -- if vim.g.colors_name then
  --   vim.cmd("hi clear")
  -- end

  vim.o.background = "dark"
  vim.o.termguicolors = true
  vim.g.colors_name = "nightfox"

  util.syntax(theme.base)
  util.syntax(theme.plugins)

  if theme.config.terminal_colors then
    util.terminal(theme)
  end
end

return util
